packages/next/src/shared/lib/router/router.ts

```diff
@@ -1776,6 +1776,10 @@ export default class Router implements BaseRouter {
           routeInfo.props.pageProps.statusCode = 500
         }
 
+        if (this.pathname === '/404' && process.env.__NEXT_HARD_NAVIGATE_404) {
```

Comment:
I'm thinking instead of hard-coding this behavior to `/404` if we provide an experimental config to allow providing custom paths to the client router filter which allows specifying paths/partial paths that should trigger hard navigations. 

x-ref: https://github.com/vercel/next.js/blob/d31b8b820646a855729ae62210e1317f15b5c550/packages/next/src/lib/create-client-router-filter.ts#L12